Add documentation re well-formedness and validation. (#1169)
usnistgov/OSCAL: docs/content/concepts/layer/validation/_index.md

---
title: Well-formed Data Formats and Valid OSCAL
weight: 50
description: Explanation of well-formedness, validation, and how they relate to OSCAL
aliases: []
suppresstopiclist: true
---

## Who is this document for?

> If you are an [OSCAL tool developer](https://pages.nist.gov/OSCAL/about/stakeholders/tooldevelopers/), you should read this document to understand OSCAL, including core terminology the OSCAL community uses in development and troubleshooting.

Reviewer: ```suggestion
If you are an [OSCAL tool developer](https://pages.nist.gov/OSCAL/about/stakeholders/tooldevelopers/), you should read this document to better understand how to use the OSCAL model. This document provides and explains some terminology the OSCAL community uses in development and troubleshooting.
```

> If you are not an OSCAL tool developer, this technical information might be useful either if you seek general understanding of OSCAL, or if as a tool *user* you encounter them in your work.

Reviewer: ```suggestion
If you are not an OSCAL tool developer, this technical information might be useful if you seek general understanding of OSCAL, or if as a tool *user* you encounter them in your work.
```

## Why are these concepts important?

> It is important that OSCAL tool developers implement software that confirm the JSON, XML, and YAML is well-formed and the [OSCAL document instances](https://pages.nist.gov/OSCAL/concepts/layer/) are valid.

Reviewer: ```suggestion
It is important that OSCAL tool developers use specialized software to confirm that JSON-, XML-, or YAML-based [OSCAL document instances](https://pages.nist.gov/OSCAL/concepts/layer/) are  well-formed and valid.
```

> Generally, both "validity" and "well-formedness" define OSCAL, such that a conformant OSCAL application can respect them, and an application that does not is *ipso facto* not an OSCAL application.

Reviewer: ```suggestion
Generally, both "well-formedness" and "validity" of OSCAL instance data determines if a conformant OSCAL application can process this data. In this way, "well-formedness" and "validity" define the boundary between what can be considered OSCAL data or not. The degree to which an application can support well-formed, valid OSCAL data defines how conformant the application is.
```

Author: NOTE: `determines`->`determine` since well-formedness and validity mean the verb should be plural. After commit, cleanup, fix that.

## What do "well-formed" and "valid" mean to the OSCAL community?

> When developing tools to read or write OSCAL, NIST or fellow community developers will ask if your data is *well-formed* and if the OSCAL document is *valid*. We define these terms below and explain how current OSCAL tools support checking well-formed data and valid OSCAL.

Reviewer: ```suggestion
When developing tools to read or write OSCAL, developers will ask if your data is *well-formed* and a *valid* OSCAL document. We define these terms below and explain how current OSCAL tools support checking well-formed data and valid OSCAL.
```

### What is well-formed data?

*Well-formed* means that your JSON, XML, YAML (whether it is OSCAL or not) follows the syntax rules of that format.

#### What OSCAL tools help you check well-formedness?

NIST's OSCAL tools do not explicitly check that JSON, XML, or YAML is well-formed. If you are a tool developer, we expect you to use your language runtime and libraries to parse JSON, XML, and YAML and check for the well-formed data. This must occur before validating an OSCAL document.

### What is a valid OSCAL document?

> *Valid* means that your OSCAL document (in a JSON, XML, or YAML already checked for *well-formedness*) follows the rules of a model. A model can be expressed using a schema format, such as [JSON Schema](https://json-schema.org/) or [XML Schema](https://www.w3.org/XML/Schema).

Reviewer: ```suggestion
*Valid* means that your JSON-, XML-, or YAML-based OSCAL document, which has already been checked for *well-formedness*, follows the structural rules of an OSCAL model. A model can be expressed using a schema format, such as [JSON Schema](https://json-schema.org/) or [XML Schema](https://www.w3.org/XML/Schema), both of which provide a means to define structural rules which can be applied in addition to the required syntax of the related format.
```

### Does OSCAL provide schemas for OSCAL models?

> The NIST OSCAL Team develops and maintains [JSON Schema](https://github.com/usnistgov/OSCAL/tree/main/json/schema) to validate OSCAL documents in JSON and YAML data formats and [XML Schema](https://github.com/usnistgov/OSCAL/tree/main/xml/schema) to validate OSCAL documents in the XML data format.

Reviewer: ```suggestion
The OSCAL program develops and maintains [JSON Schema](https://github.com/usnistgov/OSCAL/tree/main/json/schema) to validate OSCAL documents in JSON and YAML data formats, and [XML Schema](https://github.com/usnistgov/OSCAL/tree/main/xml/schema) to validate OSCAL documents in the XML data format.
```

### What tools check that an OSCAL document is valid?

There are many tools to validate documents using schema formats.

Here are some Internet resources to help you find a tool.

- [JSON Schema validator](https://json-schema.org/implementations.html#validators)
- [XML Schema validators](https://www.w3.org/XML/Schema#Tools)

**Note: These lists are not exaustive and NIST does not endorse any of these tools.**

> ### Why can OSCAL document validation fail?

Reviewer: This section seems to end abruptly. We talk about failure modes, but say nothing about how to address them. Not sure what to do about this, since we don't want to make this document much larger.

Author: I agree, so one of the remaining items with regards to this PR would be referencing to the "how to validate OSCAL?" content we agree to split into a different page, and later just a README in usnistgov/OSCAL#1170.

Would a link to this other Markdown README in the repo saying "for further explanation of how validation occurs and how to react to different failure modes, see this document" here be enough or you are thinking something more comprehensive?

Reviewer 2: This is deep b/c failure mode and appropriate remediation depends on scope of authority. Can the data be corrected, remediated or otherwise flagged, or must it be pushed back?

For one use case, e.g. editing content, to be able intervene and adjust to make invalid content into valid content is a requirement; to another (e.g., receiving data as a "blind" submission) the rule must be "no correction, only rejection".

Additionally, not all validation rules will have the same impacts, operationally, and they need to be addressed differently depending on how the data is produced.

AJ have the old standards been played yet: "when producing, be strict; when consuming, be amenable"?

Reviewer 2: Apologies, I spun off....

Author: I didn't think you spun off, I thought it was important and relevant to the discussion here. It might just go to another part of another document later.

I talked with Dave earlier in the afternoon and he recommended we purse approach 1 with additional caveat: for each one of the failure conditions in the list, add a brief sentence about the approach you would use to debug/resolve it further. After we comfortable with #1170, we will add the link in this page to that resulting GitHub link to that file.

There are multiple failure scenarios for OSCAL documents, and JSON, XML, and YAML more generally:

1. The document is well-formed, but is not conformant to the schema and errors are correctly reported.

> 2. The document is not well-formed, so validation cannot occur.

Reviewer 3: Minor observation: lack of well-formed precludes schema validation, so well-formedness should precede schema validation.

Author: Don't we cover that later int he document?

> #### What OSCAL tools help you check well-formedness?

> NIST's OSCAL tools do not explicitly check that JSON, XML, or YAML is well-formed. If you are a tool developer, we expect you to use your language runtime and libraries to parse JSON, XML, and YAML and check for the well-formed data. **This must occur before validating an OSCAL document.** (Emphasis add here by me for context.)

Author: I can see your point, but I do not say whether it is logically or temporally before (if we are addressing the snippet where I copied from).

What do you recommend we add or take away?

I only added this bit to address the question in the section header: Q: What OSCAL tools [from NIST] check well-formedness? A: NIST's OSCAL tools do not provide tools to check well-formedness, because if you are a tool developer it is on you to check that before (logically or temporally) using schemas to validate. [This implies: do not report issues back to NIST with problems in schema validation when it is easy to spot the document isn't well-formed. These are not schema issues, it should be obvious but you ought to check the doc is well-formed somewhere in your code.]

So, how can we write that better?
